find_exponential/src/main.rs

```diff
@@ -0,0 +1,197 @@
+/// it's assumed to be non-linear.
+const ACCEPTANCE_CORRELATION: f64 = 0.995;
+/// 0 / 1 / 2 / 3
+const DEBUG_LEVEL: u8 = 2;
+
+fn main() {
+    let num_cpus = num_cpus::get() * 3 / 4;
+
+    let arg = env::args().nth(1);
+    if let Some("--retest") = arg.as_ref().map(|s| s.as_str()) {
+        for pattern in io::stdin().lock().lines() {
+            let pattern = pattern.unwrap();
+            println!("retesting {:?}", pattern);
+            test(&pattern);
+        }
+        return;
+    } else {
+        fuzz(num_cpus);
+    }
+}
+
+fn fuzz(num_cpus: usize) {
+    // get all literals from pulldown-cmark's source code
+    let walkdir = WalkDir::new("../src")
+        .into_iter()
+        .filter_map(|e| e.ok())
+        .filter(|e| e.file_type().is_file())
+        .filter(|e| if let Some(ext) = e.path().extension() { ext == "rs" } else { false });
+
+    let mut literals = HashSet::new();
+    for file in walkdir {
+        if file.path() == Path::new("../src/entities.rs") {
+            continue;
+        }
+        println!("parsing {}", file.path().display());
+        let content = fs::read_to_string(file.path())
+            .expect(&format!("unable to read file {:?}", file.path()));
+        let parsed = syn::parse_file(&content)
+            .expect(&format!("unable to parse file {:?}", file.path()));
+        literals::extract_literals_from_items(&mut literals, parsed.items);
+    }
+
+    let literals: Vec<_> = literals.into_iter()
+        .filter(|lit| !lit.contains(&b'\n'))
+        .filter(|lit| !lit.is_empty())
+        .collect();
+
+    let pool = ThreadPool::new(num_cpus);
+    let mut count = -(num_cpus as isize) * BATCH_SIZE as isize;
+    let start_time = Instant::now();
+    let (tx, rx) = mpsc::channel();
+    for _ in 0..num_cpus {
+        tx.send(()).unwrap();
+    }
+    loop {
+        rx.recv().unwrap();
+        let mut literals = literals.clone();
+        let tx = tx.clone();
+        count += BATCH_SIZE as isize;
+        pool.execute(move || {
+            literals.shuffle(&mut rand::thread_rng());
+            let combs = literals.iter()
+                .combinations(COMBINATIONS)
+                .take(BATCH_SIZE);
+            for comb in combs {
+                let res = panic::catch_unwind(|| {
+                    let concatenated = comb.into_iter().flatten().cloned().collect::<Vec<_>>();
+                    let s = String::from_utf8(concatenated).unwrap();
+                    test(&s);
+                });
+                if res.is_err() {
+                    ::std::process::exit(1);
+                }
+            }
+            tx.send(()).unwrap();
+        });
+        if count > 0 {
+            println!("{:<20} throughput: {}", count, count as u64 / start_time.elapsed().as_secs());
+        }
+    }
+}
+
+fn test(pattern: &str) {
+    let (_, _, non_linear) = test_pattern(pattern, SAMPLE_SIZE, false);
+    if non_linear {
```

Comment:
Is this for retesting/ validation?

Reply:
As explained above, I perform two passes to verify if something is non-linear. The first one is the simple test, the second one the second test including outlier recalculation.

Reply:
That's a good strategy!